markdown/util.py

```diff
@@ -45,60 +44,269 @@
+        return self._set.intersection(*others)
+
+    def intersection_update(self, *others: set[str]) -> None:
+        # In-place intersection should update both list and set.
+        self._set.intersection_update(*others)
+        for element in reversed(self._list):
+            if element not in self._set:
+                self._list.remove(element)
+
+    def isdisjoint(self, other: set[str]) -> bool:
+        return self._set.isdisjoint(other)
+
+    def issubset(self, other: set[str]) -> bool:
+        return self._set.issubset(other)
+
+    def issuperset(self, other: set[str]) -> bool:
+        return self._set.issuperset(other)
+
+    def pop(self, index: int = -1, /) -> str:
+        # In-place pop should update both list and set.
+        element = self._list.pop(index)
+        self._set.remove(element)
+        return element
+
+    def remove(self, element: str) -> None:
+        # In-place removal should update both list and set.
+        self._list.remove(element)
+        self._set.remove(element)
 
     def reverse(self) -> None:
-        warnings.warn("method `reverse` will be removed in the future", DeprecationWarning)
-
-    def sort(self, /, *, key: Callable | None = None, reverse: bool = False) -> None:
-        warnings.warn("method `sort` will be removed in the future", DeprecationWarning)
-
-
-BLOCK_LEVEL_ELEMENTS: set[str] = _BlockLevelElements({
+        warnings.warn(
+            "Using block level elements as a list is deprecated, use it as a set instead.",
+            DeprecationWarning,
+        )
+        self._list.reverse()
+
+    def sort(self, /, *, key=None, reverse=False) -> None:
+        warnings.warn(
+            "Using block level elements as a list is deprecated, use it as a set instead.",
+            DeprecationWarning,
+        )
+        self._list.sort(key=key, reverse=reverse)
+
+    def symmetric_difference(self, other: set[str]) -> set[str]:
+        # User expects a set back.
+        return self._set.symmetric_difference(other)
+
+    def symmetric_difference_update(self, other: set[str]) -> None:
+        # In-place symmetric difference should update both list and set.
+        self._set.symmetric_difference_update(other)
+        for element in other:
+            if element in self._set:
+                self._list.remove(element)
+            else:
+                self._list.append(element)
+
+    def union(self, *others: set[str]) -> set[str]:
+        # User expects a set back.
+        return self._set.union(*others)
+
+    def update(self, *others: set[str]) -> None:
+        # In-place union should update both list and set.
+        self._set.update(*others)
+        for other in others:
+            for element in other:
+                if element not in self._set:
+                    self._list.append(element)
+
+
+# Type it as `set[str]` to express our intent for it to be used as such.
+BLOCK_LEVEL_ELEMENTS: set[str] = _BlockLevelElements([
```

Comment:
I can't guarantee I didn't make subtle mistakes, so it should probably need proper testing.

If this all seems too complicated, happy to try something simpler (it was implemented quickly, no problem with discarding all this :smile:).

Reply:
It should definitely get tested. Eventually it'll go away as well. There's probably ways in which people can still mess things up, but I think this attempts to provide a reasonable attempt to keep stuff from breaking. I'm open to other ideas if someone has them as well.

Reply:
Alright, I'll start writing tests :slightly_smiling_face: Just wanted initial feedback before tackling them :blush: